### PR TITLE
RDM-10564: Refactored caching for all repositories

### PR DESCRIPTION
**Pull request by rajaram-nagarajan**
### JIRA link ###

https://tools.hmcts.net/jira/browse/RDM-10564

### Change description ###

Refactored caching for all repositories

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: dependency-check-suppressions.xml [hunk 1]
```diff
@@ -35,4 +35,10 @@
 		<cve>CVE-2021-29425</cve>
 	</suppress>
 
+	<suppress until="2021-04-23">
+		<notes>Declared as False Positive on commons-io-2.2.jar
+			(pkg:maven/commons-io/commons-io@2.2, cpe:2.3:a:apache:commons_io:2.2:*:*:*:*:*:*:*)</notes>
+		<cve>CVE-2021-29425</cve>
+	</suppress>
```
Comment (MSancaktutar) [inline dependency-check-suppressions.xml:42]: Would this PR really need this in its delta?

File: charts/ccd-data-store-api/values.yaml [hunk 2]
```diff
@@ -36,9 +36,12 @@ java:
 
     DEFINITION_CACHE_MAX_IDLE_SEC: 259200
     DEFINITION_CACHE_LATEST_VERSION_TTL_SEC: 1
+    DEFINITION_CACHE_JURISDICTION_TTL_SEC: 120
+    DEFINITION_CACHE_USER_TTL_SEC: 1800
+    DEFINITION_CACHE_CASE_DETAILS_TTL_SEC: 1800
+    DEFINITION_CACHE_DRAFT_TTL_SEC: 1800
```
Comment (MSancaktutar) [inline charts/ccd-data-store-api/values.yaml:42]: where do these come from?

File: dependency-check-suppressions.xml (hunk 1 above)
Comment (rajaram-nagarajan) [reply dependency-check-suppressions.xml]: > Would this PR really need this in its delta?

This was a temporary suppression. Removed it.

File: charts/ccd-data-store-api/values.yaml (hunk 2 above)
Comment (rajaram-nagarajan) [reply charts/ccd-data-store-api/values.yaml:42]: > where do these come from?

These are all newly introduced caching variables to promote more flexibility in our caching configuration.

File: src/contractTest/resources/application.properties [hunk 3]
```diff
@@ -25,24 +25,33 @@ spring.datasource.hikari.maximum-pool-size=${DATA_STORE_DB_MAX_POOL_SIZE:16}
 spring.jpa.properties.hibernate.temp.use_jdbc_metadata_defaults = false
 pagination.page.size=25
 
+spring.liquibase.enabled=${ENABLE_DB_MIGRATE:true}
```
Comment (smathangi) [inline src/contractTest/resources/application.properties:28]: probably auto merge issue. remove these

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java [hunk 4]
```diff
@@ -86,15 +86,21 @@
...
+    @Value("${definition.cache.latest-version-ttl.secs}")
     private Integer latestVersionTTLSecs;
 
-    @Value("${definition.cache.jurisdiction-ttl}")
-    private Integer jurisdictionTTL;
+    @Value("${definition.cache.jurisdiction-ttl.secs}")
+    private Integer jurisdictionTTLSecs;
 
-    @Value("${user.cache.ttl.secs}")
+    @Value("${definition.cache.user-ttl.secs}")
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java:95]: any reason for definition prefix for these?

File: src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseRoleRepository.java [hunk 5]
```diff
@@ -17,15 +15,15 @@
...
     private final CaseRoleRepository caseRoleRepository;
-    private final Map<String, Set<String>> caseRoles = newHashMap();
 
     public CachedCaseRoleRepository(@Qualifier(DefaultCaseRoleRepository.QUALIFIER)
                                     final CaseRoleRepository caseRoleRepository) {
         this.caseRoleRepository = caseRoleRepository;
     }
 
     @Override
+    @Cacheable("caseRolesForCaseTypeCache")
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseRoleRepository.java:25]: I think normal Cacheable annotation may not work as expected with @RequestScope bean. You might need something like 
` @Cacheable(value = "caseRolesForCaseTypeCache", cacheManager = "requestScopedCacheManager")`

File: src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseUserRepository.java [hunk 6]
```diff
@@ -19,32 +19,39 @@
...
-
     public CachedCaseUserRepository(@Qualifier(DefaultCaseUserRepository.QUALIFIER)
-                                        CaseUserRepository caseUserRepository) {
+                                    final CaseUserRepository caseUserRepository) {
         this.caseUserRepository = caseUserRepository;
     }
 
     @Override
+    @Caching(evict = {
+        @CacheEvict(value = "casesForUserCache", key = "#userId", allEntries = true),
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseUserRepository.java:29]: Does it work correctly with cluster of nodes? As other nodes may update records in parallel.

**Review by smathangi — COMMENTED** [review]
Few comments. Will continue my review once we both agree on those.

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java (hunk 4 above)
Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: Reverted back the changes.

File: charts/ccd-data-store-api/values.yaml (hunk 2 above)
Comment (rajaram-nagarajan) [reply charts/ccd-data-store-api/values.yaml]: Removed all the newly introduced caching variables.

File: src/contractTest/resources/application.properties (hunk 3 above)
Comment (rajaram-nagarajan) [reply src/contractTest/resources/application.properties]: Yes, removed

File: src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseRoleRepository.java (hunk 5 above)
Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseRoleRepository.java:25]: As per our discussion, we have decided to go with the HashMap approach for @RequestScope cache beans.

File: src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseUserRepository.java (hunk 6 above)
Comment (rajaram-nagarajan) [reply src/main/java/uk/gov/hmcts/ccd/data/caseaccess/CachedCaseUserRepository.java]: It looks like the embedded cache topology will not evict the caches from cluster of instances. We might need client-server topology which has its own drawback of network communication etc. So, we have decided to stick with embedded cache topology by considering the nature of our application.

**Comment by rajaram-nagarajan** [comment]
Closing this pull request as we have to reverted back some code changes and the base is outdated

**Pull request closed without merging**